libs/libskynet/src/skylinkVerifyResolver.ts

```diff
@@ -0,0 +1,145 @@
+  }
+
+  // Now that we've confirmed all the typings are correct, pull them out into
+  // variables that will be easier to work with.
+  const dataStr = <string>proof.data;
+  const datakeyStr = <string>proof.datakey;
+  const pubkeyStr = <string>proof.publickey.key;
+  const sigStr = <string>proof.signature;
+  const revision = <bigint>proof.revision;
+
+  // Decode all of the fields. They are presented in varied types and
+  // encodings.
+  const [data, errD] = hexToBuf(dataStr);
+  if (errD !== null) {
+    return [new Uint8Array(0), addContextToErr(errD, "data is invalid hex")];
+  }
+  const [datakey, errDK] = hexToBuf(datakeyStr);
+  if (errDK !== null) {
+    return [new Uint8Array(0), addContextToErr(errDK, "datakey is invalid hex")];
+  }
+  const [pubkey, errPK] = b64ToBuf(pubkeyStr);
+  if (errPK !== null) {
+    return [new Uint8Array(0), addContextToErr(errPK, "pubkey key is invalid base64")];
+  }
+  const [sig, errS] = hexToBuf(sigStr);
+  if (errS !== null) {
+    return [new Uint8Array(0), addContextToErr(errS, "signature is invalid hex")];
+  }
+
+  // Verify that the data is a skylink - this is a proof for a resolver,
+  // which means the proof is pointing to a specific skylink.
+  if (!validSkylink(data)) {
+    return [new Uint8Array(0), "this skylink does not resolve to another skylink"];
+  }
+
+  // Verify that the combination of the datakey and the public key match
+  // the skylink.
+  const [entryID, errREID] = deriveRegistryEntryID(pubkey, datakey);
+  if (errREID !== null) {
+    return [new Uint8Array(0), addContextToErr(errREID, "proof pubkey is malformed")];
+  }
+  const linkID = skylink.slice(2, 34);
+  for (let i = 0; i < entryID.length; i++) {
+    if (entryID[i] !== linkID[i]) {
+      return [new Uint8Array(0), "proof pubkey and datakey do not match the skylink root"];
+    }
+  }
+
+  // Verify the signature.
+  if (!verifyRegistrySignature(pubkey, datakey, data, revision, sig)) {
+    return [new Uint8Array(0), "signature does not match"];
+  }
+  return [data, null];
+}
+
+// verifyResolverLinkProofs will verify a set of resolver link proofs provided
+// by a portal after performing a resolver link lookup. Each proof corresponds
+// to one level of resolution. The final value returned will be the V1 skylink
+// at the end of the chain.
+//
+// This function treats the proof as untrusted data and will verify all of the
+// fields that are provided.
+function verifyResolverLinkProofs(skylink: Uint8Array, proof: any): [Uint8Array, string | null] {
+  // Check that the proof is an array.
+  if (!Array.isArray(proof)) {
+    return [new Uint8Array(0), "provided proof is not an array: " + objAsString(proof)];
+  }
+  if (proof.length === 0) {
+    return [new Uint8Array(0), "proof array is empty"];
+  }
+
+  // Check each proof in the chain, returning the final skylink.
+  for (let i = 0; i < proof.length; i++) {
+    let errVRLP;
+    [skylink, errVRLP] = verifyResolverLinkProof(skylink, proof[i]);
```

Comment:
why are we overwriting skylink function argument on each iteration? I would define `let rootSkylink` instead and only assign it once we reach the last iteration to be explicit about intent

Reply:
you have to pass in the new skylink to the proof verifier. This loop basically forms a daisy chain of verifications.